File: python/csrc/sampling.cpp
#include "common.h"

namespace loom_kernels::torch_adapter {

void check_greedy_sample_logprobs_shape(const Tensor& logits) {
  STD_TORCH_CHECK(logits.dim() == 2 && logits.size(0) > 0 && logits.size(1) > 0,
              "Loom greedy sampling logits must be non-empty rank-2");
  STD_TORCH_CHECK(logits.size(0) <= std::numeric_limits<uint32_t>::max() &&
                  logits.size(1) <= std::numeric_limits<int32_t>::max(),
              "Loom greedy sampling shape exceeds the CUDA ABI");
}

void check_greedy_sample_logprobs_contract(const Tensor& logits) {
  check_greedy_sample_logprobs_shape(logits);
  STD_TORCH_CHECK(logits.is_cuda(), "Loom greedy sampling logits must be CUDA");
  STD_TORCH_CHECK(logits.scalar_type() == ScalarType::Float ||
                  logits.scalar_type() == ScalarType::Half ||
                  logits.scalar_type() == ScalarType::BFloat16,
              "Loom greedy sampling supports F32, FP16, and BF16 logits");
  STD_TORCH_CHECK(logits.stride(1) == 1 && logits.stride(0) >= logits.size(1),
              "Loom greedy sampling logits require unit vocabulary stride "
              "and non-overlapping positive row stride");
}

std::tuple<Tensor, Tensor, Tensor>
launch_greedy_sample_logprobs(const Tensor& logits) {
  const auto rows = static_cast<uint32_t>(logits.size(0));
  const auto vocab_size = static_cast<uint32_t>(logits.size(1));
  const auto row_stride = static_cast<uint64_t>(logits.stride(0));
  const auto logits_elements = storage_span_elements(logits);
  const auto output_elements = static_cast<uint64_t>(logits.size(0));
  Tensor token_ids = new_empty(logits, {logits.size(0)}, ScalarType::Int);
  Tensor logprobs = new_empty(logits, {logits.size(0)}, ScalarType::Float);
  Tensor ranks = new_empty(logits, {logits.size(0)}, ScalarType::Long);

  const CudaDeviceGuard device_guard(logits.device());
  const auto stream = current_cuda_stream(logits.device().index());
  const int status = loom_cuda_bridge_greedy_sample_logprobs(
      bridge_dtype(logits), logits.const_data_ptr(), logits_elements,
      token_ids.mutable_data_ptr<int32_t>(), output_elements,
      logprobs.mutable_data_ptr<float>(), output_elements,
      ranks.mutable_data_ptr<int64_t>(),
      output_elements, rows, vocab_size, row_stride, stream.stream());
  check_bridge_status(status, "greedy-sampling");
  return {token_ids, logprobs, ranks};
}

std::tuple<Tensor, Tensor, Tensor> greedy_sample_logprobs(
    const Tensor& logits) {
  check_greedy_sample_logprobs_contract(logits);
  return launch_greedy_sample_logprobs(logits);
}

std::tuple<Tensor, Tensor, Tensor> greedy_sample_logprobs_meta(
    const Tensor& logits) {
  check_greedy_sample_logprobs_shape(logits);
  return {
      new_empty(logits, {logits.size(0)}, ScalarType::Int),
      new_empty(logits, {logits.size(0)}, ScalarType::Float),
      new_empty(logits, {logits.size(0)}, ScalarType::Long),
  };
}

void check_selected_token_logprobs_shape(const Tensor& logits,
                                         const Tensor& token_ids) {
  check_greedy_sample_logprobs_shape(logits);
  STD_TORCH_CHECK(token_ids.dim() == 1 && token_ids.size(0) == logits.size(0),
              "Loom selected token IDs must contain one value per logits row");
}

void check_selected_token_logprobs_contract(const Tensor& logits,
                                            const Tensor& token_ids) {
  check_greedy_sample_logprobs_contract(logits);
  check_selected_token_logprobs_shape(logits, token_ids);
  STD_TORCH_CHECK(token_ids.device() == logits.device(),
              "Loom selected token IDs and logits must share a CUDA device");
  STD_TORCH_CHECK(token_ids.scalar_type() == ScalarType::Long,
              "Loom selected token IDs must be int64");
  STD_TORCH_CHECK(token_ids.is_contiguous(),
              "Loom selected token IDs must be contiguous");
}

std::tuple<Tensor, Tensor> launch_selected_token_logprobs(
    const Tensor& logits, const Tensor& token_ids) {
  const auto rows = static_cast<uint32_t>(logits.size(0));
  const auto vocab_size = static_cast<uint32_t>(logits.size(1));
  const auto row_stride = static_cast<uint64_t>(logits.stride(0));
  Tensor logprobs = new_empty(logits, {logits.size(0)}, ScalarType::Float);
  Tensor ranks = new_empty(logits, {logits.size(0)}, ScalarType::Long);

  const CudaDeviceGuard device_guard(logits.device());
  const auto stream = current_cuda_stream(logits.device().index());
  const auto output_elements = static_cast<uint64_t>(logits.size(0));
  const int status = loom_cuda_bridge_selected_token_logprobs(
      bridge_dtype(logits), logits.const_data_ptr(),
      storage_span_elements(logits), token_ids.const_data_ptr<int64_t>(),
      static_cast<uint64_t>(token_ids.numel()),
      logprobs.mutable_data_ptr<float>(), output_elements,
      ranks.mutable_data_ptr<int64_t>(),
      output_elements, rows, vocab_size, row_stride, stream.stream());
  check_bridge_status(status, "selected-token logprob");
  return {logprobs, ranks};
}

std::tuple<Tensor, Tensor> selected_token_logprobs(
    const Tensor& logits, const Tensor& token_ids) {
  check_selected_token_logprobs_contract(logits, token_ids);
  return launch_selected_token_logprobs(logits, token_ids);
}

std::tuple<Tensor, Tensor> selected_token_logprobs_meta(
    const Tensor& logits, const Tensor& token_ids) {
  check_selected_token_logprobs_shape(logits, token_ids);
  return {
      new_empty(logits, {logits.size(0)}, ScalarType::Float),
      new_empty(logits, {logits.size(0)}, ScalarType::Long),
  };
}

void check_topk_sampled_logprobs_shape(
    const Tensor& logits, const Tensor& sampled_token_ids, int64_t top_k) {
  check_selected_token_logprobs_shape(logits, sampled_token_ids);
  const int64_t maximum = std::min<int64_t>(logits.size(1), 32);
  STD_TORCH_CHECK(
      top_k > 0 && top_k <= maximum,
      "Loom top-k sampled logprobs require 1 <= top_k <= ", maximum,
      "; got ", top_k);
}

void check_topk_sampled_logprobs_contract(
    const Tensor& logits, const Tensor& sampled_token_ids, int64_t top_k) {
  check_selected_token_logprobs_contract(logits, sampled_token_ids);
  check_topk_sampled_logprobs_shape(logits, sampled_token_ids, top_k);
}

std::tuple<Tensor, Tensor, Tensor> launch_topk_sampled_logprobs(
    const Tensor& logits, const Tensor& sampled_token_ids, int64_t top_k) {
  const auto rows = static_cast<uint32_t>(logits.size(0));
  const auto vocab_size = static_cast<uint32_t>(logits.size(1));
  const auto row_stride = static_cast<uint64_t>(logits.stride(0));
  const int64_t output_width = top_k + 1;
  const uint64_t output_elements =
      static_cast<uint64_t>(logits.size(0)) *
      static_cast<uint64_t>(output_width);
  uint64_t workspace_bytes = 0;
  const int workspace_status =
      loom_cuda_bridge_topk_sampled_logprobs_workspace_size(
          rows, vocab_size, static_cast<uint32_t>(top_k), &workspace_bytes);
  check_bridge_status(workspace_status, "top-k workspace query");
  STD_TORCH_CHECK(
      workspace_bytes <=
          static_cast<uint64_t>(std::numeric_limits<int64_t>::max()),
      "Loom top-k workspace exceeds the PyTorch shape ABI");
  Tensor output_token_ids =
      new_empty(logits, {logits.size(0), output_width}, ScalarType::Int);
  Tensor output_logprobs =
      new_empty(logits, {logits.size(0), output_width}, ScalarType::Float);
  Tensor sampled_token_ranks =
      new_empty(logits, {logits.size(0)}, ScalarType::Long);
  Tensor workspace = new_empty(
      logits, {static_cast<int64_t>(workspace_bytes)}, ScalarType::Byte);

  const CudaDeviceGuard device_guard(logits.device());
  const auto stream = current_cuda_stream(logits.device().index());
  const int status = loom_cuda_bridge_topk_sampled_logprobs(
      bridge_dtype(logits), logits.const_data_ptr(),
      storage_span_elements(logits),
      sampled_token_ids.const_data_ptr<int64_t>(),
      static_cast<uint64_t>(sampled_token_ids.numel()),
      output_token_ids.mutable_data_ptr<int32_t>(), output_elements,
      output_logprobs.mutable_data_ptr<float>(), output_elements,
      sampled_token_ranks.mutable_data_ptr<int64_t>(),
      static_cast<uint64_t>(sampled_token_ranks.numel()),
      workspace.mutable_data_ptr<uint8_t>(),
      static_cast<uint64_t>(workspace.numel()), rows, vocab_size,
      static_cast<uint32_t>(top_k), row_stride, stream.stream());
  check_bridge_status(status, "top-k sampled logprobs");
  return {output_token_ids, output_logprobs, sampled_token_ranks};
}

std::tuple<Tensor, Tensor, Tensor> topk_sampled_logprobs(
    const Tensor& logits, const Tensor& sampled_token_ids, int64_t top_k) {
  check_topk_sampled_logprobs_contract(logits, sampled_token_ids, top_k);
  return launch_topk_sampled_logprobs(logits, sampled_token_ids, top_k);
}

std::tuple<Tensor, Tensor, Tensor> topk_sampled_logprobs_meta(
    const Tensor& logits, const Tensor& sampled_token_ids, int64_t top_k) {
  check_topk_sampled_logprobs_shape(logits, sampled_token_ids, top_k);
  const int64_t output_width = top_k + 1;
  return {
      new_empty(logits, {logits.size(0), output_width}, ScalarType::Int),
      new_empty(logits, {logits.size(0), output_width}, ScalarType::Float),
      new_empty(logits, {logits.size(0)}, ScalarType::Long),
  };
}

uint64_t required_token_penalty_workspace(int64_t prompt_tokens,
                                          int64_t output_tokens) {
  STD_TORCH_CHECK(prompt_tokens > 0 && output_tokens > 0,
              "Loom token-penalty history widths must be positive");
  const uint64_t prompt = static_cast<uint64_t>(prompt_tokens);
  const uint64_t output = static_cast<uint64_t>(output_tokens);
  STD_TORCH_CHECK(
      prompt <= (std::numeric_limits<uint64_t>::max() - output) &&
          prompt + output <=
              std::numeric_limits<uint32_t>::max() / 2ULL,
      "Loom token-penalty workspace exceeds the CUDA ABI");
  uint64_t required = 2 * (prompt + output);
  --required;
  required |= required >> 1;
  required |= required >> 2;
  required |= required >> 4;
  required |= required >> 8;
  required |= required >> 16;
  required |= required >> 32;
  return required + 1;
}

void check_token_penalties_shape(
    const Tensor& logits, const Tensor& prompt_token_ids,
    const Tensor& output_token_ids, const Tensor& presence_penalties,
    const Tensor& frequency_penalties, const Tensor& repetition_penalties,
    const Tensor& workspace) {
  STD_TORCH_CHECK(logits.dim() == 2 && logits.size(0) > 0 && logits.size(1) > 0,
              "Loom token-penalty logits must be non-empty rank-2");
  const int64_t rows = logits.size(0);
  STD_TORCH_CHECK(prompt_token_ids.dim() == 2 &&
                  prompt_token_ids.size(0) == rows &&
                  prompt_token_ids.size(1) > 0,
              "Loom prompt token IDs must be [rows, positive width]");
  STD_TORCH_CHECK(output_token_ids.dim() == 2 &&
                  output_token_ids.size(0) == rows &&
                  output_token_ids.size(1) > 0,
              "Loom output token IDs must be [rows, positive width]");
  for (const auto* penalty :
       {&presence_penalties, &frequency_penalties, &repetition_penalties}) {
    STD_TORCH_CHECK(penalty->dim() == 1 && penalty->size(0) == rows,
                "Loom token penalties must contain one value per logits row");
  }
  STD_TORCH_CHECK(workspace.dim() == 2 && workspace.size(0) == rows &&
                  workspace.size(1) > 0,
              "Loom token-penalty workspace must be [rows, positive capacity]");
  const uint64_t capacity = static_cast<uint64_t>(workspace.size(1));
  const uint64_t required = required_token_penalty_workspace(
      prompt_token_ids.size(1), output_token_ids.size(1));
  STD_TORCH_CHECK((capacity & (capacity - 1)) == 0 && capacity >= required,
              "Loom token-penalty workspace capacity must be a power of two "
              "and at least twice the combined history width; required ",
              required, ", got ", capacity);
  STD_TORCH_CHECK(
      rows <= std::numeric_limits<int32_t>::max() &&
          logits.size(1) <= std::numeric_limits<int32_t>::max() &&
          prompt_token_ids.size(1) <= std::numeric_limits<uint32_t>::max() &&
          output_token_ids.size(1) <= std::numeric_limits<int32_t>::max() &&
          workspace.size(1) <= std::numeric_limits<uint32_t>::max(),
      "Loom token-penalty shape exceeds the CUDA ABI");
}

void check_token_penalties_contract(
    const Tensor& logits, const Tensor& prompt_token_ids,
    const Tensor& output_token_ids, const Tensor& presence_penalties,
    const Tensor& frequency_penalties, const Tensor& repetition_penalties,
    const Tensor& workspace) {
  check_token_penalties_shape(
      logits, prompt_token_ids, output_token_ids, presence_penalties,
      frequency_penalties, repetition_penalties, workspace);
  STD_TORCH_CHECK(logits.is_cuda() && prompt_token_ids.is_cuda() &&
                  output_token_ids.is_cuda() && presence_penalties.is_cuda() &&
                  frequency_penalties.is_cuda() &&
                  repetition_penalties.is_cuda() && workspace.is_cuda(),
              "Loom token penalties require CUDA tensors");
  const auto device = logits.device();
  STD_TORCH_CHECK(prompt_token_ids.device() == device &&
                  output_token_ids.device() == device &&
                  presence_penalties.device() == device &&
                  frequency_penalties.device() == device &&
                  repetition_penalties.device() == device &&
                  workspace.device() == device,
              "Loom token-penalty tensors must share one CUDA device");
  STD_TORCH_CHECK(logits.scalar_type() == ScalarType::Float,
              "Loom token-penalty logits must be float32");
  STD_TORCH_CHECK(prompt_token_ids.scalar_type() == ScalarType::Long &&
                  output_token_ids.scalar_type() == ScalarType::Long,
              "Loom token IDs must be int64");
  STD_TORCH_CHECK(presence_penalties.scalar_type() == ScalarType::Float &&
                  frequency_penalties.scalar_type() == ScalarType::Float &&
                  repetition_penalties.scalar_type() == ScalarType::Float,
              "Loom token penalties must be float32");
  STD_TORCH_CHECK(workspace.scalar_type() == ScalarType::Long,
              "Loom token-penalty workspace must be int64");
  STD_TORCH_CHECK(logits.stride(1) == 1 &&
                  logits.stride(0) >= logits.size(1) &&
                  prompt_token_ids.stride(1) == 1 &&
                  prompt_token_ids.stride(0) >= prompt_token_ids.size(1) &&
                  output_token_ids.stride(1) == 1 &&
                  output_token_ids.stride(0) >= output_token_ids.size(1) &&
                  workspace.stride(1) == 1 &&
                  workspace.stride(0) >= workspace.size(1),
              "Loom token-penalty matrices require unit inner stride and "
              "non-overlapping positive row strides");
  STD_TORCH_CHECK(presence_penalties.is_contiguous() &&
                  frequency_penalties.is_contiguous() &&
                  repetition_penalties.is_contiguous(),
              "Loom token-penalty vectors must be contiguous");
}

void apply_token_penalties_(
    Tensor logits, const Tensor& prompt_token_ids,
    const Tensor& output_token_ids, const Tensor& presence_penalties,
    const Tensor& frequency_penalties, const Tensor& repetition_penalties,
    Tensor workspace) {
  check_token_penalties_contract(
      logits, prompt_token_ids, output_token_ids, presence_penalties,
      frequency_penalties, repetition_penalties, workspace);
  const auto rows = static_cast<uint32_t>(logits.size(0));
  const auto vocab_size = static_cast<uint32_t>(logits.size(1));
  const auto prompt_tokens = static_cast<uint32_t>(prompt_token_ids.size(1));
  const auto output_tokens = static_cast<uint32_t>(output_token_ids.size(1));
  const auto workspace_capacity = static_cast<uint32_t>(workspace.size(1));
  const CudaDeviceGuard device_guard(logits.device());
  const auto stream = current_cuda_stream(logits.device().index());
  const int status = loom_cuda_bridge_apply_token_penalties(
      logits.mutable_data_ptr<float>(), storage_span_elements(logits),
      prompt_token_ids.const_data_ptr<int64_t>(),
      storage_span_elements(prompt_token_ids),
      output_token_ids.const_data_ptr<int64_t>(),
      storage_span_elements(output_token_ids),
      presence_penalties.const_data_ptr<float>(),
      static_cast<uint64_t>(presence_penalties.numel()),
      frequency_penalties.const_data_ptr<float>(),
      static_cast<uint64_t>(frequency_penalties.numel()),
      repetition_penalties.const_data_ptr<float>(),
      static_cast<uint64_t>(repetition_penalties.numel()),
      reinterpret_cast<uint64_t*>(workspace.mutable_data_ptr<int64_t>()),
      storage_span_elements(workspace), rows, vocab_size, prompt_tokens,
      output_tokens, workspace_capacity,
      static_cast<uint64_t>(logits.stride(0)),
      static_cast<uint64_t>(prompt_token_ids.stride(0)),
      static_cast<uint64_t>(output_token_ids.stride(0)),
      static_cast<uint64_t>(workspace.stride(0)), stream.stream());
  check_bridge_status(status, "token penalties");
}

void apply_token_penalties_meta(
    Tensor logits, const Tensor& prompt_token_ids,
    const Tensor& output_token_ids, const Tensor& presence_penalties,
    const Tensor& frequency_penalties, const Tensor& repetition_penalties,
    Tensor workspace) {
  check_token_penalties_shape(
      logits, prompt_token_ids, output_token_ids, presence_penalties,
      frequency_penalties, repetition_penalties, workspace);
}


}  // namespace loom_kernels::torch_adapter

STABLE_TORCH_LIBRARY_IMPL(loom_kernels, CUDA, library) {
  library.impl(
      "greedy_sample_logprobs",
      TORCH_BOX(&loom_kernels::torch_adapter::greedy_sample_logprobs));
  library.impl(
      "selected_token_logprobs",
      TORCH_BOX(&loom_kernels::torch_adapter::selected_token_logprobs));
  library.impl(
      "topk_sampled_logprobs",
      TORCH_BOX(&loom_kernels::torch_adapter::topk_sampled_logprobs));
  library.impl(
      "apply_token_penalties_",
      TORCH_BOX(&loom_kernels::torch_adapter::apply_token_penalties_));
}

STABLE_TORCH_LIBRARY_IMPL(loom_kernels, Meta, library) {
  library.impl(
      "greedy_sample_logprobs",
      TORCH_BOX(&loom_kernels::torch_adapter::greedy_sample_logprobs_meta));
  library.impl(
      "selected_token_logprobs",
      TORCH_BOX(&loom_kernels::torch_adapter::selected_token_logprobs_meta));
  library.impl(
      "topk_sampled_logprobs",
      TORCH_BOX(&loom_kernels::torch_adapter::topk_sampled_logprobs_meta));
  library.impl(
      "apply_token_penalties_",
      TORCH_BOX(&loom_kernels::torch_adapter::apply_token_penalties_meta));
}

File: python/src/loom_kernels/ops/sampling.py
"""Sampling predicates, validation, and public PyTorch APIs."""

from __future__ import annotations

import torch

from .._torch_dispatch import (
    _apply_token_penalties,
    _greedy_sample_logprobs,
    _selected_token_logprobs,
    _top_k_filter,
    _topk_sampled_logprobs,
)
from ._common import _DTYPE_NAMES


def supports_greedy_sample_logprobs(logits: torch.Tensor) -> bool:
    """Return whether logits match the deterministic greedy CUDA boundary."""
    return bool(
        logits.device.type == "cuda"
        and logits.dtype in _DTYPE_NAMES
        and logits.dim() == 2
        and logits.shape[0] > 0
        and logits.shape[1] > 0
        and logits.shape[0] <= 0xFFFF_FFFF
        and logits.shape[1] <= 0x7FFF_FFFF
        and logits.stride(1) == 1
        and logits.stride(0) >= logits.shape[1]
        and not logits.requires_grad
    )


def supports_selected_token_logprobs(
    logits: torch.Tensor, token_ids: torch.Tensor
) -> bool:
    """Return whether one selected token per logits row can be normalized."""
    return bool(
        supports_greedy_sample_logprobs(logits)
        and token_ids.device == logits.device
        and token_ids.dtype == torch.int64
        and token_ids.dim() == 1
        and token_ids.shape[0] == logits.shape[0]
        and token_ids.is_contiguous()
        and not token_ids.requires_grad
    )


def supports_topk_sampled_logprobs(
    logits: torch.Tensor,
    sampled_token_ids: torch.Tensor,
    top_k: int,
) -> bool:
    """Return whether sampled-token plus top-k logprobs can be fused."""
    return bool(
        supports_selected_token_logprobs(logits, sampled_token_ids)
        and isinstance(top_k, int)
        and not isinstance(top_k, bool)
        and 1 <= top_k <= min(logits.shape[1], 32)
    )


def supports_top_k_filter(
    logits: torch.Tensor,
    top_ks: torch.Tensor,
) -> bool:
    """Return whether tensors match the exact in-place top-k boundary."""
    return bool(
        supports_greedy_sample_logprobs(logits)
        and top_ks.device == logits.device
        and top_ks.dtype == torch.int32
        and top_ks.dim() == 1
        and top_ks.shape[0] == logits.shape[0]
        and top_ks.is_contiguous()
        and not top_ks.requires_grad
    )


def token_penalties_workspace_capacity(
    prompt_tokens: int,
    output_tokens: int,
) -> int:
    """Return packed int64 hash slots required per logits row."""
    if prompt_tokens <= 0 or output_tokens <= 0:
        raise ValueError("token-penalty history widths must be positive")
    return 1 << (2 * (prompt_tokens + output_tokens) - 1).bit_length()


def supports_apply_token_penalties(
    logits: torch.Tensor,
    prompt_token_ids: torch.Tensor,
    output_token_ids: torch.Tensor,
    presence_penalties: torch.Tensor,
    frequency_penalties: torch.Tensor,
    repetition_penalties: torch.Tensor,
    workspace: torch.Tensor,
) -> bool:
    """Return whether tensors satisfy the sparse F32 penalty boundary."""
    if not (
        logits.device.type == "cuda"
        and logits.dtype == torch.float32
        and logits.dim() == 2
        and logits.shape[0] > 0
        and logits.shape[1] > 0
        and logits.shape[0] <= 0x7FFF_FFFF
        and logits.shape[1] <= 0x7FFF_FFFF
        and logits.stride(1) == 1
        and logits.stride(0) >= logits.shape[1]
        and not logits.requires_grad
    ):
        return False
    rows = logits.shape[0]
    matrices = (prompt_token_ids, output_token_ids)
    if not all(
        tensor.device == logits.device
        and tensor.dtype == torch.int64
        and tensor.dim() == 2
        and tensor.shape[0] == rows
        and tensor.shape[1] > 0
        and tensor.stride(1) == 1
        and tensor.stride(0) >= tensor.shape[1]
        and not tensor.requires_grad
        for tensor in matrices
    ):
        return False
    penalties = (
        presence_penalties,
        frequency_penalties,
        repetition_penalties,
    )
    if not all(
        tensor.device == logits.device
        and tensor.dtype == torch.float32
        and tensor.dim() == 1
        and tensor.shape[0] == rows
        and tensor.is_contiguous()
        and not tensor.requires_grad
        for tensor in penalties
    ):
        return False
    required = token_penalties_workspace_capacity(
        prompt_token_ids.shape[1], output_token_ids.shape[1]
    )
    return bool(
        workspace.device == logits.device
        and workspace.dtype == torch.int64
        and workspace.dim() == 2
        and workspace.shape[0] == rows
        and workspace.shape[1] >= required
        and workspace.shape[1] <= 0xFFFF_FFFF
        and (workspace.shape[1] & (workspace.shape[1] - 1)) == 0
        and workspace.stride(1) == 1
        and workspace.stride(0) >= workspace.shape[1]
        and not workspace.requires_grad
    )


def _validate_greedy_sample_logits(
    logits: torch.Tensor,
) -> None:
    if not supports_greedy_sample_logprobs(logits):
        raise ValueError(
            "Loom greedy sampling requires finite, non-empty rank-2 "
            "F32/FP16/BF16 CUDA logits with unit vocabulary stride, "
            "non-overlapping rows, and no gradients"
        )


def _validate_selected_token_logprobs(
    logits: torch.Tensor,
    token_ids: torch.Tensor,
) -> None:
    if not supports_selected_token_logprobs(logits, token_ids):
        raise ValueError(
            "Loom selected-token logprobs require finite, non-empty rank-2 "
            "F32/FP16/BF16 CUDA logits with unit vocabulary stride and one "
            "same-device contiguous int64 token ID per row; token IDs must "
            "be in vocabulary range"
        )


def _validate_topk_sampled_logprobs(
    logits: torch.Tensor,
    sampled_token_ids: torch.Tensor,
    top_k: int,
) -> None:
    if not supports_topk_sampled_logprobs(logits, sampled_token_ids, top_k):
        maximum = min(logits.shape[1], 32) if logits.dim() == 2 else 32
        raise ValueError(
            "Loom top-k sampled logprobs require inference-only, non-empty "
            "rank-2 F32/FP16/BF16 CUDA logits with unit vocabulary stride; "
            "one same-device contiguous int64 sampled token ID per row; and "
            f"1 <= top_k <= {maximum}"
        )


def _validate_top_k_filter(
    logits: torch.Tensor,
    top_ks: torch.Tensor,
) -> None:
    if not supports_top_k_filter(logits, top_ks):
        raise ValueError(
            "Loom top-k filtering requires inference-only, non-empty rank-2 "
            "F32/FP16/BF16 CUDA logits with unit vocabulary stride and one "
            "same-device contiguous int32 top-k value per row"
        )


def _validate_apply_token_penalties(
    logits: torch.Tensor,
    prompt_token_ids: torch.Tensor,
    output_token_ids: torch.Tensor,
    presence_penalties: torch.Tensor,
    frequency_penalties: torch.Tensor,
    repetition_penalties: torch.Tensor,
    workspace: torch.Tensor,
) -> None:
    if not supports_apply_token_penalties(
        logits,
        prompt_token_ids,
        output_token_ids,
        presence_penalties,
        frequency_penalties,
        repetition_penalties,
        workspace,
    ):
        raise ValueError(
            "Loom token penalties require inference-only F32 rank-2 CUDA "
            "logits; same-device row-major int64 prompt/output IDs; one "
            "contiguous F32 value per row for presence, frequency, and "
            "repetition; and a same-device int64 power-of-two workspace with "
            "at least twice the combined padded history width"
        )


def greedy_sample_logprobs(
    logits: torch.Tensor,
) -> tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
    """Return greedy IDs, sampled logprobs, and vLLM-compatible tie ranks."""
    _validate_greedy_sample_logits(logits)
    return _greedy_sample_logprobs(logits)


def selected_token_logprobs(
    logits: torch.Tensor,
    token_ids: torch.Tensor,
) -> tuple[torch.Tensor, torch.Tensor]:
    """Return normalized logprobs and ranks for one selected token per row."""
    _validate_selected_token_logprobs(logits, token_ids)
    return _selected_token_logprobs(logits, token_ids)


def top_k_filter_(
    logits: torch.Tensor,
    top_ks: torch.Tensor,
) -> torch.Tensor:
    """Apply exact per-row top-k thresholds in place, preserving ties."""
    _validate_top_k_filter(logits, top_ks)
    _top_k_filter(logits, top_ks)
    return logits


def topk_sampled_logprobs(
    logits: torch.Tensor,
    sampled_token_ids: torch.Tensor,
    top_k: int,
) -> tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
    """Return sampled token, deterministic top-k logprobs, and sampled ranks."""
    _validate_topk_sampled_logprobs(logits, sampled_token_ids, top_k)
    return _topk_sampled_logprobs(logits, sampled_token_ids, top_k)


def apply_token_penalties_(
    logits: torch.Tensor,
    prompt_token_ids: torch.Tensor,
    output_token_ids: torch.Tensor,
    presence_penalties: torch.Tensor,
    frequency_penalties: torch.Tensor,
    repetition_penalties: torch.Tensor,
    workspace: torch.Tensor,
) -> torch.Tensor:
    """Apply sparse vLLM-compatible penalties in place with caller workspace."""
    _validate_apply_token_penalties(
        logits,
        prompt_token_ids,
        output_token_ids,
        presence_penalties,
        frequency_penalties,
        repetition_penalties,
        workspace,
    )
    _apply_token_penalties(
        logits,
        prompt_token_ids,
        output_token_ids,
        presence_penalties,
        frequency_penalties,
        repetition_penalties,
        workspace,
    )
    return logits
